### PR TITLE
Introduce test directives for including and excluding files

### PR DESCRIPTION
**Pull request by bmsantos**
Added test directives to:
- Include desired globs (defaults to "**/*Test.class". Note that ".class" can be omitted.)
- Exclude desired globs (no default)
- Updated GlobSpec to do both include and exclude filtering. This allowed to quite a bit of code cleaning and simplification.
- Other minor improvements

### CODE REVIEW

File: modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/FileSpec.kt [hunk 1]
```diff
@@ -3,36 +3,96 @@ package com.beust.kobalt
...
 import java.io.File
 import java.nio.file.*
+import java.nio.file.FileVisitResult.CONTINUE
 import java.nio.file.attribute.BasicFileAttributes
+import java.util.*
+import kotlin.collections.*
 
 sealed class IFileSpec {
     abstract fun toFiles(directory: String): List<File>
+    abstract fun toFile(file: String): List<File>
```
Comment (cbeust) [inline modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/FileSpec.kt]: It's called `toFile` but it returns a list of files... that's a bit confusing.

File: modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/FileSpec.kt [hunk 2]
```diff
@@ -3,36 +3,96 @@ package com.beust.kobalt
...
 
-        override public fun toString() = spec
+        override public fun toFile(file: String): List<File> {
+
+            val result = arrayListOf<File>()
+            val includeMatchers = prepareMatchers(includeSpec.toTypedArray())
+            val excludeMatchers = prepareMatchers(excludeSpec.toTypedArray())
+
+            val rel = Paths.get(file)
+            excludeMatchers.forEach {
```
Comment (cbeust) [inline modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/FileSpec.kt]: BIt of duplication with the code above, we can probably extract some common code here (doesn't need to be in this PR).

File: modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/TestDirective.kt [hunk 3]
```diff
@@ -11,6 +13,17 @@ class TestConfig(val project: Project) {
...
+
+    fun includes(vararg arg: String) {
+        project.testIncludes.apply {
+            clear()
+            addAll(arg)
+        }
+    }
+
+    fun excludes(vararg arg: String) {
+        project.testExcludes.addAll(arg)
```
Comment (cbeust) [inline modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/TestDirective.kt]: Not symmetric with the code above, probably needs a `clear()` as well? Let's make both codes similar.

File: modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/api/Project.kt [hunk 4]
```diff
@@ -15,18 +19,20 @@ open class Project(
         @Directive open var artifactId: String? = null,
         @Directive open var packaging: String? = null,
         @Directive open var dependencies: Dependencies? = null,
-        @Directive open var sourceSuffix : String = "",
-        @Directive open var description : String = "",
-        @Directive open var scm : Scm? = null,
```
Comment (cbeust) [inline modules/kobalt-plugin-api/src/main/kotlin/com/beust/kobalt/api/Project.kt]: FYI, I try to put cosmetic changes (reformatting, reindenting, etc...) in separate PR so they don't obscure the main PR. No big deal for now.

**Comment by cbeust** [comment]
Hi @bmsantos,

Thanks for the PR!

Looking good overall but most of it is just space deletion/inserting, reformatting and import moving around, which is why it looks so big. I'd like to have a PR with just the code and I'm happy to look at another one with the cosmetic stuff.

If you don't feel like doing that, it's okay, I'll do it myself based on your code.

Let me know!

**Comment by cbeust** [comment]
I ended up extracting the significant code from your PR and commit it:

https://github.com/cbeust/kobalt/commit/eb2937098d353079e9bf63ae26d43893f8f89a97

Let me know what you think!

**Comment by cbeust** [comment]
Oh and I updated the doc too:  http://beust.com/kobalt/plug-ins/index.html#tests

**Comment by bmsantos** [comment]
Looks really good.

Sorry for the late response but just had some free time now.
I was hoping to have another stab in trying to simplify the GlobSpec code but you beat me at it. :disappointed: :smile:

**Pull request closed without merging**